RFC.md

```diff
@@ -0,0 +1,413 @@
+- The need to support mixtures of layer types where some might require conditional inputs, some might not.
+
+**TLDR** Layers need to support a primary input and mask (current inputs) and an optional conditional input and mask. This would support any standard variation of TransformerBlock.
+
+**Note on Attention** In the diagram above, it should be noted that all three attention implementations (red, green, and blue) use the same attention module, Multi Headed Attention. Most of the implementation complexity lie inside this module, and it's only getting more complex [#1193](https://github.com/pytorch/torchtune/pull/1193). The only difference between them is what inputs are provided as k, q, v, and what mask is provided. All of that to say, to support cross attention and self attention, the attention module doesn't change but the transformer block itself. See Figure 2, Deep Fusion, for an overview of some popular layer variations for multimodal models [ref](https://arxiv.org/pdf/2405.17927).
+## Proposal
+**High Level List:**
+- Extend the TransformerDecoder forward signature to include `encoder_input` and `encoder_mask`. This sets a standard signature for all layers to follow so they can all be mixed and matched and stacked together.
+- Modify TransformerDecoder to accept a list of layers instead of a layer and and a count. This allows us to support models where only some layers are conditional and other are not.
+- Implement two initial layers `TransformerSelfAttentionLayer` and `TransformerCrossAttentionLayer`. `TransformerSelfAttentionLayer` would replace `TransformerDecoderLayer`. These would just be two possible layers, but the design would support more variations.
+- Generalize `CausalSelfAttention` work in all three scenarios above (red, green, blue) and rename to `GroupedQueryAttention`. This includes changing the default mask behavior to not always be causal.
+- **Extra**:
+	- Add TanhGate module to support gated attention [ref](https://paperswithcode.com/method/attention-gate)
+	- Modify `TransformerDecoder` to output it's hidden layers and make `model.output` optional to support the changes proposed [#1017](https://github.com/pytorch/torchtune/issues/1017) and make only one BC breaking change.
+	- Implement legacy checkpoint conversion to support old checkpoints
+
+### TransformerDecoder
+TransformerDecoder only highlighting the changes, current module [here](https://github.com/pytorch/torchtune/blob/0057fe7cf83e14f0b62538a8d4d20719e0a88639/torchtune/modules/transformer.py#L99). Discussion in the comments.
+
+```python
+class TransformerDecoder(nn.Module):
+    def __init__(
+        self,
+        *,
+        ...
+        # as opposed to layer and num_layers
+        layers: List[nn.Module],
+        ...
+        # output is generalized and hidden_idx is added to address #1017
+        output: Callable
+        hidden_idx: Optional[List[int]] = None
```

Comment:
wth is this

Reply:
List of hidden layers to return. It could also be called `hidden_layers`, `output_layers`, `output_indices`, etc.

Reply:
Ya could be named better. Personally I just like `return_hidden_states` as a `Union[bool, List[int]]` that we convert to a `List[bool]` (or `List[int]`) on the backend; that way we can still easily support the common case of returning all hidden layers pretty easily.

Reply:
75% noob question: should `hidden_idx` be a property of `TransformerDecoder`? Or are users expected to have to re-initialise if they want different states?

Reply:
@SalmanMohammadi I was wondering about this too. I anticipate it may help with compile to have a fixed set of layers that are deterministically returned on each forward pass (though haven't actually tested this so please take with a huge grain of salt). On the other hand, `output_hidden_states` is usually defined in HF as part of the forward signature so it may be what users are accustomed to. 

But there are also certain things that may necessitate this being a static property of the class. E.g. if we are using hidden states as part of a fusion down the line and we concatenate them along the sequence dimension before passing to an MLP or something, we would actually need to know the number of hidden states to infer the MLP dim. I am not sure if this is something @pbontrager is trying to solve for, but if so we would probably need to do it this way.

Reply:
¿por qué no los dos¿

Couldn't we override in the forwad pass like we do `self.causal mask` and `mask`? Preferring `output_hidden_states: Optional[...] = None` over `self._output_hidden_states`?

Reply:
+ 1 to the name being `output_hidden_states` and consideration for having this be defined on the main module, but possible to override in the forward pass if needed. 

The latter could be overkill for this refactor and I would settle for a code comment mentioning that this is possible to enable if necessary.

Reply:
I don't know why this would ever be dynamically changing in the forward pass. I'd rather just expose it in the builder function if we needed too.  I'll use `output_hidden_states`.

Do people prefer the output `return output, hidden` or 
```
output = output if not hidden else hidden + [output]
return output
```
?